packages/@expo/cli/src/localModules/generation.ts

```diff
@@ -0,0 +1,362 @@
+import { getConfig } from '@expo/config';
+import { getPbxproj } from '@expo/config-plugins/build/ios/utils/Xcodeproj';
+import Server from '@expo/metro/metro/Server';
+import type MetroServer from '@expo/metro/metro/Server';
+import fs from 'fs';
+import path from 'path';
+
+import { ensureDotExpoProjectDirectoryInitialized } from '../start/project/dotExpo';
+import { getRouterDirectoryModuleIdWithManifest } from '../start/server/metro/router';
+
+export interface ModuleGenerationArguments {
+  projectRoot: string;
+  metro: Server | null;
+}
+
+const nativeExtensions = ['.kt', '.swift'];
+
+function getMirrorDirectories(projectRoot: string): {
+  localModulesModulesPath: string;
+  localModulesTypesPath: string;
+} {
+  const dotExpoDir = ensureDotExpoProjectDirectoryInitialized(projectRoot);
+  const localModulesPath = path.resolve(dotExpoDir, './localModules/');
+
+  const localModulesModulesPath = path.resolve(localModulesPath, 'modules');
+  const localModulesTypesPath = path.resolve(localModulesPath, 'types');
+
+  return {
+    localModulesModulesPath,
+    localModulesTypesPath,
+  };
+}
+
+function createFreshMirrorDirectories(projectRoot: string) {
+  const { localModulesModulesPath, localModulesTypesPath } = getMirrorDirectories(projectRoot);
+
+  // make sure the directories exist so we can remove them.
+  if (!fs.existsSync(localModulesModulesPath)) {
+    fs.mkdirSync(localModulesModulesPath, { recursive: true });
+  }
+  if (!fs.existsSync(localModulesTypesPath)) {
+    fs.mkdirSync(localModulesTypesPath, { recursive: true });
+  }
+}
+
+function trimExtension(fileName: string) {
+  return fileName.substring(0, fileName.lastIndexOf('.'));
+}
+
+function typesAndLocalModulePathsForFile(projectRoot: string, absoluteFilePath: string) {
+  const { localModulesModulesPath, localModulesTypesPath } = getMirrorDirectories(projectRoot);
+  const splitPath = absoluteFilePath.split('/');
+  const fileName = splitPath.at(-1);
+  if (!fileName)
+    throw new Error("In local modules we shouldn't watch files other than .kt and .swift");
+  const moduleName = trimExtension(fileName);
```

Comment:
what if you have a file like `SomeFile.someStr.ext` or `my-module.ts` – should we add validation/coercion for proper modulenames?